src/magic_folder/snapshot.py

```diff
@@ -131,13 +162,96 @@ def create_local_author_from_config(config, name=None):
     )
 
 
+def create_author(name, verify_key):
+    """
+    :param name: arbitrary name for this author
+
+    :param verify_key: a NaCl VerifyKey instance
+
+    :returns: a RemoteAuthor instance.
+    """
+    if not isinstance(verify_key, VerifyKey):
+        raise ValueError("verify_key must be a nacl.signing.VerifyKey")
+
+    return RemoteAuthor(
+        name=name,
+        verify_key=verify_key,
+    )
+
+
+def create_author_from_json(data):
+    """
+    :returns: a RemoteAuthor instance from the given data (which
+       would usually come from RemoteAuthor.to_json())
+    """
+    permitted_keys = required_keys = ["name", "verify_key"]
+    for k in data.keys():
+        if k not in permitted_keys:
+            raise ValueError(
+                u"Unknown RemoteAuthor key '{}'".format(k)
+            )
+    for k in required_keys:
+        if k not in data:
+            raise ValueError(
+                u"RemoteAuthor requires '{}' key".format(k)
+            )
+    verify_key = VerifyKey(data["verify_key"], encoder=Base64Encoder)
+    return create_author(data["name"], verify_key)
+
+
+def sign_snapshot(local_author, snapshot, content_capability):
+    """
+    Signs the given snapshot with the given key
+
+    :param SigningKey signing_key: the key to sign the data with
+
+    :param LocalSnapshot snapshot: snapshot to sign
+
+    :param str content_capability: the Tahoe immutable capability of
+        the actual snapshot data.
+
+    :returns: bytes representing the signature or exception on
+        error.
+    """
+    # XXX what do we sign? Should we hash it first? Ask our cryptographers
+    data_to_sign = (
+        u"{content_capability}\n"
+        u"{name}\n"
+    ).format(
+        content_capability=content_capability,
+        name=snapshot.name,
+    )
+    return local_author.signing_key.sign(data_to_sign.encode("utf8"))
+
+
+def verify_snapshot_signature(remote_author, alleged_signature, content_capability, snapshot_name):
+    """
+    Verify the given snapshot.
+
+    :returns: True on success or exception otherwise
+    """
+    # See comments about "data_to_sign" in sign_snapshot
+    data_to_verify = (
+        u"{content_capability}\n"
+        u"{name}\n"
+    ).format(
+        content_capability=content_capability,
+        name=snapshot_name,
+    )
+    return remote_author.verify_key.verify(
+        data_to_verify.encode("utf8"),
+        alleged_signature,
+    )
+
+
 @attr.s
 class LocalSnapshot(object):
     name = attr.ib()
     author = attr.ib()
     metadata = attr.ib()
     content_path = attr.ib()  # full filesystem path to our stashed contents
     parents_local = attr.ib()  # LocalSnapshot instances
+    parents_remote = attr.ib(default=attr.Factory(list))  # capability-strings
```

Comment:
If they weren't separate, anyone looking at `parents` would have to decide if they have a `LocalSnapshot` instance or a `RemoteSnapshot` instance. And, they're really quite different sorts of things (I can't imagine any case where you wouldn't either want "the list of `LocalSnapshot`s" or "the list of `RemoteSnapshot`s").

At least, that's what I now imagine I was thinking then.

Reply:
(On the flip side, what would be the point of having just one `parents` or similar list?)

Reply:
It's difficult to reason about without having some application code around, I think.  I'm trying to imagine what that will look like now.

Ah, but I suppose I don't have to imagine because it is `write_snapshot_to_tahoe`.  I thought it might be part of a later PR adding some higher-level uploader service thing.

I think the idea of a work queue has been mentioned a couple times here and there.  `write_snapshot_to_tahoe` doesn't seem like it fits with this idea particularly well since it recursively processes `parents_local`.  If a snapshot close to being a leaf ends up early in any queue then it's going to force processing of a bunch of other snapshots that might also be in the queue.  If we wanted to have roughly-similar-sized work units processed from a queue it would probably be preferable not to have the recursive processing and instead require that all parents are already remote - then the queue would have to be populated to respect this constraint.  That looks a bit tricky given the current schema since we essentially have to load every snapshot, parse the JSON, and then inspect the result to determine if it needs upload and if it has only remote parents.  On the other hand, populating the queue is going to be approximately that hard anyway without a different schema ... unless `local_snapshots` is meant to be emptied out as the snapshots are uploaded?

So ... I'm not really sure how these pieces are intended to fit together.  Perhaps we could work that out in a some more detail (or just communicate what has already been worked out)?


It's probably primarily the uploader, right?  It wants to get a `LocalSnapshot` that needs to be uploaded and upload it.  To upload it, all of the parents need to have been uploaded.  All of the un-uploaded parents are in `parents_local` so it can visit each of those and do the same thing to them.  Does it ... mutate the `LocalSnapshot` as it goes, taking things from `parents_local` and adding them to `parents_remote`?  That doesn't sound so hot.  Really it needs to be interacting with the database as part of this process because that's the persistent source of truth.  If it uploads anything it needs to put